From: Paul Bone <pbone@mozilla.com>
Date: Thu, 23 Nov 2023 04:53:22 +0000
Subject: Bug 1866025 - Check if PHC is initialised on all entrypoints
 r=glandium

PHC initialisation can bail out on systems without an expected statically
assumed page size.  When this happens Firefox can crash if we don't check
for it on all PHC entrypoints including SetPHCStatus().

Thanks Janne Grunau for the initial patch.

Differential Revision: https://phabricator.services.mozilla.com/D194458
---
 memory/build/PHC.cpp | 27 ++++++++++++++++++++++-----
 1 file changed, 22 insertions(+), 5 deletions(-)

diff --git a/memory/build/PHC.cpp b/memory/build/PHC.cpp
index a1bfb4a..d20b2ab 100644
--- a/memory/build/PHC.cpp
+++ b/memory/build/PHC.cpp
@@ -1696,6 +1696,10 @@ inline void* MozJemallocPHC::moz_arena_memalign(arena_id_t aArenaId,
 namespace mozilla::phc {
 
 bool IsPHCAllocation(const void* aPtr, AddrInfo* aOut) {
+  if (!maybe_init()) {
+    return false;
+  }
+
   PtrKind pk = gConst->PtrKind(aPtr);
   if (pk.IsNothing()) {
     return false;
@@ -1760,6 +1764,11 @@ bool IsPHCEnabledOnCurrentThread() {
 }
 
 void PHCMemoryUsage(MemoryUsage& aMemoryUsage) {
+  if (!maybe_init()) {
+    aMemoryUsage = MemoryUsage();
+    return;
+  }
+
   aMemoryUsage.mMetadataBytes = metadata_size();
   if (gMut) {
     MutexAutoLock lock(GMut::sMutex);
@@ -1770,15 +1779,23 @@ void PHCMemoryUsage(MemoryUsage& aMemoryUsage) {
 }
 
 void GetPHCStats(PHCStats& aStats) {
-  if (gMut) {
-    MutexAutoLock lock(GMut::sMutex);
-
-    aStats = gMut->GetPageStats(lock);
+  if (!maybe_init()) {
+    aStats = PHCStats();
+    return;
   }
+
+  MutexAutoLock lock(GMut::sMutex);
+
+  aStats = gMut->GetPageStats(lock);
 }
 
 // Enable or Disable PHC at runtime.  If PHC is disabled it will still trap
 // bad uses of previous allocations, but won't track any new allocations.
-void SetPHCState(PHCState aState) { gMut->SetState(aState); }
+void SetPHCState(PHCState aState) {
+  if (!maybe_init()) {
+    return;
+  }
 
+  gMut->SetState(aState);
+}
 }  // namespace mozilla::phc
